Add shared cache for resolvers

brianwcook opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Resolvers are awesome but can be noisy. In high traffic environments we have seen rate limiting both from registries when using the bundle resolver and from git forges when using the git resolver. This PR adds caching to the resolvers (except the cluster resolver). It is only enabled by default when 'safe' - meaning that the object was fetched by some hash / digest. Otherwise, it can be turned on with a resolver parameter.

> [!NOTE]  
 This PR is a work in progress posted for initial thoughts and feedback. It works for some cases already (git resolver) - others, maybe not.



<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Remote resolver results for bundle and git resolvers which use immutable references will be cached by default for 5 minutes. User can opt-in to caching non-immutable references or disable caching entirely. The default caching mode (auto) should be safe and entirely transparent for existing users.
```

twoGiants commented:
/assign

brianwcook commented:
/retest

vdemeester commented:
/retest

twoGiants reviewed (commented):
Hi @brianwcook ! Thank you for this great contribution 😸  👍 

The PR status is **open** and I started reviewing but it looks like its still work in progress if I am not mistaken. I thought to publish the comments I have so far anyway, I hope it's ok and hope they are helpful.

I think this PR would benefit from a few e2e tests.

Another consideration is to maybe extend caching to the cluster resolver. I had a conversation with a user on Slack last week and he _might_ be running into rate limiting imposed by k8 when fetching many (100+) tasks. Here is the issue #8889, but it has not all the details. The [thread in slack](https://tektoncd.slack.com/archives/CJ62C1555/p1752698665708749) has more details.

twoGiants commented on pkg/remoteresolution/resolver/git/resolver.go:
```diff
@@ -127,13 +185,24 @@ func (r *Resolver) Resolve(ctx context.Context, req *v1beta1.ResolutionRequestSp
+		var resource resolutionframework.ResolvedResource
 		if params[git.UrlParam] != "" {
-			return g.ResolveGitClone(ctx)
+			resource, err = g.ResolveGitClone(ctx)
```
If you put `ResolveGitClone` and `ResolveAPIGit` as function properties on the struct, as you did in GitResolver you could overwrite them here for testing and testing of this method becomes easier.

twoGiants commented on pkg/remoteresolution/resolver/bundle/resolver.go:
```diff
@@ -77,11 +95,82 @@ func (r *Resolver) Validate(ctx context.Context, req *v1beta1.ResolutionRequestS
+
+			// Check cache first
+			if cached, ok := cache.GetGlobalCache().Get(cacheKey); ok {
+				if resource, ok := cached.(resolutionframework.ResolvedResource); ok {
```
Just food for thought: does it makes sense to find a way to enrich `ResolvedResource` somehow with an annotation that it was cached? This would make unit/e2e tests easier. Wdys?

brianwcook commented:
Thank you @twoGiants for the thorough review!

I've addressed the following points from your feedback:

- Guard patterns - Restructured both resolvers with early returns
- Cache as resolver property - Added cache field and updated initialization  
- Simplified cache usage - Enhanced cache interface to handle key generation internally
- Function properties for testing - Added injectable functions for easier unit testing
- Inlined useCache variables - Removed unnecessary local variables
- Type safety - Enhanced cache to be more specific to ResolvedResource
- Code structure - Renamed existing cache to secretCache and improved error handling
- Cache interface - Added methods that encapsulate key generation

In addition I added cache support for the cluster resolver and added more tests, including integration / e2e tests.

twoGiants commented:
> I've addressed the following points from your feedback:
> ...

Thank you for addressing @brianwcook! :+1: 

It looks like you forgot to push the latest changes.

brianwcook commented:
whoops 😔 
pushed

brianwcook commented:
Actually, I was too hasty in declaring victory on the clusterResolver caching. The current implementation is broken. In order to properly support cluster tasks I think we have the following options:

1) use mutating webhook to add  checksum to cluster task objects and use that as the cache key. It won't help inline task specs unless we add more webhooks.
2) accept that cluster tasks are mutable and be defensive by requiring explicit cache: true. (auto-mode is always false)
3) don't offer caching for clusterResolver. am i missing anything?

If you like #2, I think we can add that in this PR. If you like #1 better, I would prefer to remove the clusterResolver code from this PR and offer a followup PR to add that, as **I am being affected** by the lack of git and bundle resolver and would like to see that get merged.

twoGiants commented:
> Actually, I was too hasty in declaring victory on the clusterResolver caching. The current implementation is broken. In order to properly support cluster tasks I think we have the following options:

Thanks for checking it out already. I prefer to first validate if the k8 rate limiting is an actual issue before we add caching to solve it. But great that you explored it and already found what will be the issue implementing it.

> I would prefer to remove the clusterResolver code from this PR and offer a followup PR to add that, as I am being affected by the lack of git and bundle resolver and would like to see that get merged.

Agree, lets do git and bundle first. 

Request another review from me when I can take another look! Looking forward to it :smile_cat: :+1:

brianwcook commented:
@twoGiants do you have any hint about why the tests fail only on that subset of kuberenetes matrix test?

twoGiants commented:
> @twoGiants do you have any hint about why the tests fail only on that subset of kuberenetes matrix test?

Hi @brianwcook! The e2e cache tests are failing with all k8 versions and the beta feature gates enabled. It looks like the current cache implementation does not play well in beta mode. For example for the e2e test `TestCacheErrorHandling` it fails because the TaskRun does not finish, [here is the error code](https://github.com/tektoncd/pipeline/pull/8825/files#diff-c0c825ed4e67d524e53445967d32f92b8027ecd413c23b3ee7489cf4f861b4f5R1060).

It's impossible to debug the e2e tests remotely. The output is interleaved because the tests run in parallel and there are 50k+ lines of log output. But you can at least find which tests fail by using the search on the top right with "--- FAIL" (see screenshot below).

I would isolate one test e.g. `TestBundleResolverCache` or  `TestCacheErrorHandling` and run it locally in KinD and the [beta feature gates enabled](https://tekton.dev/docs/pipelines/additional-configs/#beta-features), then you should be able to corner it more easy with more log outputs or debugging ([vscode setup](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#debugging)) enabled.

Here is how you set up [e2e tests locally](https://github.com/tektoncd/pipeline/blob/main/test/README.md#setup), what you need there are only this env vars setup: `KO_DOCKER_REPO` and `SYSTEM_NAMESPACE`. Then run it with [this command.](https://github.com/tektoncd/pipeline/blob/main/test/README.md#running-specific-test-cases)

<img width="1374" height="606" alt="Screenshot From 2025-07-26 10-36-01" src="https://github.com/user-attachments/assets/f717c98d-36e5-477e-ab53-fc2e05261ba5" />
<img width="1295" height="420" alt="Screenshot From 2025-07-26 10-32-59" src="https://github.com/user-attachments/assets/5dd518e6-2ffe-42db-8315-620bc2a914d8" />

aThorp96 commented on pkg/remoteresolution/cache/cache.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+// InitializeLogger initializes the logger for the cache using the provided context
+func (c *ResolverCache) InitializeLogger(ctx context.Context) {
+	if c.logger == nil {
+		c.logger = logging.FromContext(ctx)
+	}
+}
```
Bug: 
This will leak state. The ResolverCache is a global object and should not include the context from the resolvers `Resolve()` call. I suggest providing the context to every call like `cache.GetGlobalCache().Get(ctx, key)` or by allowing the context to decorate the global cache along these lines:
```suggestion
func  WithLogger(logger *zap.SugaredLogger) c *ResolverCache {
	return ResolverCache{logger: logger, cache: GetGlobalCache().cache}
}
```

aThorp96 reviewed (commented):
Agreed with @twoGiants

A refactor would be prudent here. I would also prefer if the `CacheMode` is not unique per resolver. IMO the logic for when a resolver should cache should be mostly handled at the framework level, with the exception that each resolver be responsible for defining what "auto" means in their context. 

There are also some bugs with the way the cache singleton is handles

brianwcook commented:
@aThorp96 @twoGiants @vdemeester, thanks for all of your early comments, I think it definitely helped to get me going in the proper direction. I think this is ready for your review now.  🎉

twoGiants commented on pkg/remoteresolution/cache/injection/cache.go:
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2024 The Tekton Authors
```
resolved

twoGiants commented on pkg/remoteresolution/cache/cache_test.go:
```diff
@@ -0,0 +1,433 @@
+		name         string
+		resolverType string
+		params       []pipelinev1.Param
+		wantErr      bool
```
resolved

twoGiants commented on pkg/remoteresolution/cache/injection/cache.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+// Key is used as the key for associating information with a context.Context.
+type Key struct{}
```
resolved

twoGiants commented on pkg/remoteresolution/resolver/cluster/resolver_test.go:
resolved

waveywaves commented:
I will move all my commits to the other PR

twoGiants commented on pkg/remoteresolution/resolver/cluster/resolver_integration_test.go:
```diff
@@ -0,0 +1,1815 @@
+	}
+}
+
+func TestResolverCacheModeConstants(t *testing.T) {
```
resolved

twoGiants commented on pkg/remoteresolution/resolver/cluster/resolver.go:
```diff
@@ -38,59 +40,121 @@ const (
+	CacheModeAlways = framework.CacheModeAlways
+	CacheModeNever  = framework.CacheModeNever
+	CacheModeAuto   = framework.CacheModeAuto
+	CacheParam      = framework.CacheParam
```
resolved
